hydromt/data_sources/data_source.py

```diff
@@ -0,0 +1,53 @@
+"""Abstract DataSource class."""
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from hydromt import DataCatalog
+from hydromt.metadata_resolvers.resolver_plugin import RESOLVERS
+
+
+class DataSource(BaseModel):
+    """
+    A DataSource is a parsed section of a DataCatalog.
+
+    The DataSource, specific for a data type within HydroMT, is responsible for
+    validating the input from the DataCatalog, to
+    ensure the workflow fails as early as possible. A DataSource has information on
+    the driver that the data should be read with, and is responsible for initializing
+    this driver.
+    """
+
+    @classmethod
+    def from_catalog(
+        cls,
+        catalog: DataCatalog,
+        key: str,
+        provider: str | None = None,
+        version: str | None = None,
+    ) -> "DataSource":
+        """Create Data source from DataCatalog."""
+        return cls.model_validate(
+            catalog.get_source(key, provider, version).update("name", key)
+        )
+
+    def _validate_metadata_resolver(cls, v: Any):
+        assert isinstance(v, str), "metadata_resolver should be string."
+        assert v in RESOLVERS, f"unknown MetaDataResolver: '{v}'."
+        RESOLVERS.get(v)
+
+    name: str
+    attrs: dict[str, Any] = Field(default_factory=dict)
+    version: str | None = Field(default=None)
+    provider: str | None = Field(default=None)
+    driver: str
+    metadata_resolver: str
```

Comment:
can be an item for later, but seeing the code in the resolver plugin file I think this sould be an enum or literal.

Reply:
Later on people should be able to write plugins where they register new drivers or resolvers. Then we should identify them with a string, or dynamically build an enum. What do you prefer?

Reply:
Forgot about that. Honestly if we build enumbs dinamically they mostly lose their usefulness to me, so strings are fine.